WalletWasabi/Hwi2/Parsers/HwiParser.cs

```diff
@@ -0,0 +1,377 @@
+			else if (err == "Not initialized")
+			{
+				code = HwiErrorCode.DeviceNotInitialized;
+			}
+
+			if (code.HasValue)
+			{
+				error = new HwiException(code.Value, err);
+			}
+			else if (err != "")
+			{
+				error = new HwiException(HwiErrorCode.UnknownError, err);
+			}
+			else if (successToken != null && successToken.Value<bool>() == false)
+			{
+				error = new HwiException(HwiErrorCode.UnknownError, "");
+			}
+
+			return error != null;
+		}
+
+		public static bool TryParseErrorCode(JToken codeToken, out HwiErrorCode code)
+		{
+			code = default;
+
+			if (codeToken is null)
+			{
+				return false;
+			}
+
+			try
+			{
+				var codeInt = codeToken.Value<int>();
+				if (Enum.IsDefined(typeof(HwiErrorCode), codeInt))
+				{
+					code = (HwiErrorCode)codeInt;
+					return true;
+				}
+			}
+			catch
+			{
+				return false;
+			}
+
+			return false;
+		}
+
+		public static bool TryParseHardwareWalletVendor(JToken token, out HardwareWalletVendors vendor)
+		{
+			vendor = default;
+
+			if (token is null)
+			{
+				return false;
+			}
+
+			try
+			{
+				var typeString = token.Value<string>();
+				if (Enum.TryParse(typeString, ignoreCase: true, out HardwareWalletVendors t))
+				{
+					vendor = t;
+					return true;
+				}
+			}
+			catch
+			{
+				return false;
+			}
+
+			return false;
+		}
+
+		public static IEnumerable<HwiEnumerateEntry> ParseHwiEnumerateResponse(string responseString)
+		{
+			var jarr = JArray.Parse(responseString);
+
+			var response = new List<HwiEnumerateEntry>();
+			foreach (JObject json in jarr)
+			{
+				var hwiEntry = ParseHwiEnumerateEntry(json);
+				response.Add(hwiEntry);
+			}
+
+			return response;
+		}
+
+		public static ExtPubKey ParseExtPubKey(string json)
+		{
+			if (JsonHelpers.TryParseJToken(json, out JToken token))
+			{
+				var extPubKeyString = token["xpub"]?.ToString()?.Trim() ?? null;
+				var extPubKey = NBitcoinHelpers.BetterParseExtPubKey(extPubKeyString);
+				return extPubKey;
+			}
+			else
+			{
+				throw new FormatException($"Could not parse extpubkey: {json}");
+			}
+		}
+
+		public static BitcoinAddress ParseAddress(string json, Network network)
+		{
+			// HWI does not support regtest, so the parsing would fail here.
+			if (network == Network.RegTest)
+			{
+				network = Network.TestNet;
+			}
+
+			if (JsonHelpers.TryParseJToken(json, out JToken token))
+			{
+				var addressString = token["address"]?.ToString()?.Trim() ?? null;
+				var address = BitcoinAddress.Create(addressString, network);
+				return address;
+			}
+			else
+			{
+				throw new FormatException($"Could not parse address: {json}");
+			}
+		}
+
+		public static PSBT ParsePsbt(string json, Network network)
+		{
+			// HWI does not support regtest, so the parsing would fail here.
+			if (network == Network.RegTest)
+			{
+				network = Network.TestNet;
+			}
+
+			if (JsonHelpers.TryParseJToken(json, out JToken token))
+			{
+				var psbtString = token["psbt"]?.ToString()?.Trim() ?? null;
```

Comment:
This doesn't look good. The first that I see is that `ToString()?` but ToString cannot return `null`. in case `token["psbt"]` is `null`, psbtString will be also null and it will be passed to the `PSBT.Parse` method what is not really necessary.

Reply:
Check my solution, should be fine.